contracts/StakeManager.sol

```diff
@@ -138,27 +138,30 @@ contract StakeManager is Ownable {
     /**
      * @notice Process epoch if it has ended
      */
-    modifier finalizeEpoch() {
-        if (block.timestamp >= epochEnd() && address(migration) == address(0)) {
+    function finalizeEpoch() private {
+        if (address(migration) != address(0)) {
+            return;
+        }
+        Epoch storage thisEpoch = epochs[currentEpoch];
+        if (block.timestamp >= thisEpoch.startTime + EPOCH_SIZE) {
             uint256 expiredMP = stakeRewardEstimate.getExpiredMP(currentEpoch);
             if (expiredMP > 0) {
                 totalMPPerEpoch -= expiredMP;
                 stakeRewardEstimate.deleteExpiredMP(currentEpoch);
             }
-            epochs[currentEpoch].estimatedMP = totalMPPerEpoch - currentEpochTotalExpiredMP;
+            thisEpoch.estimatedMP = totalMPPerEpoch - currentEpochTotalExpiredMP;
             delete currentEpochTotalExpiredMP;
-            pendingMPToBeMinted += epochs[currentEpoch].estimatedMP;
+            pendingMPToBeMinted += thisEpoch.estimatedMP;
 
             //finalize current epoch
-            epochs[currentEpoch].epochReward = epochReward();
-            epochs[currentEpoch].totalSupply = totalSupply();
-            pendingReward += epochs[currentEpoch].epochReward;
+            thisEpoch.epochReward = epochReward();
+            thisEpoch.totalSupply = totalSupply();
+            pendingReward += thisEpoch.epochReward;
 
             //create new epoch
             currentEpoch++;
-            epochs[currentEpoch].startTime = block.timestamp;
+            epochs[currentEpoch].startTime = thisEpoch.startTime + EPOCH_SIZE;
```

Comment:
Soo.. this looks a bit fishy, but maybe I'm wrong.

Given that we set `epochs[currentEpoch].startTime` to `thisEpoch.startTime + EPOCH_SIZE` this skews the time at which staking is happening no?

Here's what I mean:

1. Protocol is at, say, epoch 2 which should finalize today
2. Nobody interacts with the contract for 3 more months
3. Someone wants to stake, and finalizes epoch here.
4. As part of finalization, the new epoch (3) is introduced at epoch(2).endTime (which actually lies 2.5 months in the past

Now the actual staking happens and from the perspective of the protocol, it looks like the account has staked 2.5 months ago (when epoch 3 has started retroactively)

Does this make sense?

Reply:
Yes, it makes sense, it's a good catch.

The other PR #129 should fix this issue, because it will process all epochs until it reached the latest epoch possible, making so that user will be at, in your example, epoch 3.

Reply:
@3esmit thanks for the quick feedback!

I fear this is not solving the issue fundamentally :(

https://github.com/logos-co/staking/pull/129#discussion_r1771062147